packages/utils/crypto/src/encryption.ts

```diff
@@ -1,19 +1,218 @@
 import { AccessError } from "./errors.js";
 import sodium from "libsodium-wrappers";
 import { X25519Keypair, X25519PublicKey, X25519SecretKey } from "./x25519.js";
-import { Ed25519Keypair, Ed25519PublicKey } from "./ed25519.js";
+import { Ed25519PublicKey } from "./ed25519.js";
 import { randomBytes } from "./random.js";
-import { Keychain } from "./keychain.js";
+import { sha256 } from "./hash.js";
+
+export type PublicKeyEncryptionParameters = {
+	type?: "publicKey";
+	receiverPublicKeys: (X25519PublicKey | Ed25519PublicKey)[];
+};
+
+export type SymmetricKeyEncryptionParameters = {
+	type?: "hash";
+};
+/* 
+export type NoExchange = {
+	type: 'none'
+};
+ */
+
+export type KeyExchangeOptions =
+	| PublicKeyEncryptionParameters
+	| SymmetricKeyEncryptionParameters;
+
+type EncryptReturnValue<
+	T,
+	Parameters extends KeyExchangeOptions
+> = EncryptedThing<T, EnvelopeFromParameter<Parameters>>;
+
+type CipherWithEnvelope<E = PublicKeyEnvelope | HashedKeyEnvelope> = {
+	cipher: Uint8Array;
+	nonce: Uint8Array;
+	envelope: E;
+};
+
+type SymmetricKeys = Uint8Array;
+type PublicKeyEncryptionKeys = X25519Keypair;
+
+function isAsymmetriEncryptionParameters(
+	parameters: KeyExchangeOptions
+): parameters is PublicKeyEncryptionParameters {
+	return (
+		(parameters as PublicKeyEncryptionParameters).receiverPublicKeys != null
+	);
+}
+function isAsymmetricEncryptionKeys(
+	parameters: PublicKeyEncryptionKeys | SymmetricKeys
+): parameters is PublicKeyEncryptionKeys {
+	return (parameters as PublicKeyEncryptionKeys) instanceof X25519Keypair;
+}
+
+type EnvelopeFromParameter<Parameters extends KeyExchangeOptions> =
+	Parameters extends PublicKeyEncryptionParameters
+		? PublicKeyEnvelope
+		: HashedKeyEnvelope;
+
+type EncryptProvide<Parameters extends KeyExchangeOptions> = (
+	bytes: Uint8Array,
+	parameters: Parameters
+) => Promise<CipherWithEnvelope<EnvelopeFromParameter<Parameters>>>;
+
+export const createLocalEncryptProvider = <
+	K extends PublicKeyEncryptionKeys | SymmetricKeys,
+	Parameters extends KeyExchangeOptions = K extends PublicKeyEncryptionKeys
+		? PublicKeyEncryptionParameters
+		: SymmetricKeyEncryptionParameters
+>(
+	keys: K
+) => {
+	return async (
+		bytes: Uint8Array,
+		parameters: Parameters
+	): Promise<CipherWithEnvelope<EnvelopeFromParameter<Parameters>>> => {
+		const nonce = randomBytes(NONCE_LENGTH); // crypto random is faster than sodim random
+		if (
+			isAsymmetriEncryptionParameters(parameters) &&
+			isAsymmetricEncryptionKeys(keys)
+		) {
+			const epheremalKey = sodium.crypto_secretbox_keygen();
+			const cipher = sodium.crypto_secretbox_easy(bytes, nonce, epheremalKey);
+			const { receiverPublicKeys } = parameters;
+			const receiverX25519PublicKeys = await Promise.all(
+				receiverPublicKeys.map((key) => {
+					if (key instanceof Ed25519PublicKey) {
+						return X25519PublicKey.from(key);
+					}
+					return key;
+				})
+			);
+
+			const ks = receiverX25519PublicKeys.map((receiverPublicKey) => {
```

Comment:
I think we should also encrypt for the sender - so take `key.publicKey` into the recipientPublicKeys array?

Reply:
I went into this question a few months ago. But I realize that a valid case which needs to be coverred is when the sender generated a throwaway X25519Keypair and just want to send a message to someone and never read it themselves


then it will be redudant to also encrypt for the sender. 

BUT 

when you do encryption in the document store, the sender is adding themselves to the receivers array because beofre the encrypt function is called because it is very likely that when you put documents into a document store you also want to read the documents later.